### ISSUE TITLE
Validation layers return function pointers for non-existant entry points

### ISSUE THREAD

**Issue by baldurk** [KhronosGroup/Vulkan-ValidationLayers#579]
This is very closely related to #304. With the same repro (included again below) the validation layers don't crash, but they call into a do-nothing stub function.

This means that an application can query for e.g. `vkGetBufferMemoryRequirements2` and get back a valid function pointer, but if the driver doesn't implement it calling the function pointer will do nothing. The layers should either return `NULL` for a function that isn't implemented or enabled, or else return a valid function pointer that does call into the driver.

Specifically this can cause problems with RenderDoc and the validation layers both being enabled. An application requesting `vkGetBufferMemoryRequirements2KHR` from RenderDoc will get a valid pointer, but RenderDoc will enumerate both `vkGetBufferMemoryRequirements2KHR` and `vkGetBufferMemoryRequirements2` onwards, and will prefer calling into the core function. Since they must be aliases this is valid, and if the core function isn't supported it should be `NULL` and the KHR variant will be called. With the current validation layers this means that even if the application calls `KHR` variant RenderDoc will redirect it onwards to the core function which ends up in a stub.

To repro either use a 1.0 driver that supports a KHR extension, or hack the validation layers in `vk_dispatch_table_helper.h` to force `NULL` instead of `gpa(device, "vkGetBufferMemoryRequirements2");`. Then run the cube demo with this patch:

```patch
diff --git a/cube.c b/cube_mod.c
index 8edbd65..74d44d4 100644
--- a/cube.c
+++ b/cube_mod.c
@@ -3175,6 +3175,8 @@ static void demo_init_vk(struct demo *demo) {
         free(device_extensions);
     }

+    demo->extension_names[demo->enabled_extension_count++] = VK_KHR_GET_MEMORY_REQUIREMENTS_2_EXTENSION_NAME;
+
     if (!swapchainExtFound) {
         ERR_EXIT("vkEnumerateDeviceExtensionProperties failed to find the " VK_KHR_SWAPCHAIN_EXTENSION_NAME
                  " extension.\n\nDo you have a compatible Vulkan installable client driver (ICD) installed?\n"
@@ -3241,6 +3243,8 @@ static void demo_init_vk(struct demo *demo) {
 }

 static void demo_create_device(struct demo *demo) {
+    VkBufferMemoryRequirementsInfo2 buf_req;
+    VkBufferCreateInfo buf_info;
     VkResult U_ASSERT_ONLY err;
     float queue_priorities[1] = {0.0};
     VkDeviceQueueCreateInfo queues[2];
@@ -3273,6 +3277,32 @@ static void demo_create_device(struct demo *demo) {
     }
     err = vkCreateDevice(demo->gpu, &device, NULL, &demo->device);
     assert(!err);
+
+    memset(&buf_info, 0, sizeof(buf_info));
+    buf_info.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO;
+    buf_info.usage = VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT;
+    buf_info.size = 1024;
+
+    memset(&buf_req, 0, sizeof(buf_req));
+    buf_req.sType = VK_STRUCTURE_TYPE_BUFFER_MEMORY_REQUIREMENTS_INFO_2;
+    vkCreateBuffer(demo->device, &buf_info, NULL, &buf_req.buffer);
+
+    PFN_vkGetBufferMemoryRequirements2KHR khr_variant =
+      (PFN_vkGetBufferMemoryRequirements2KHR)vkGetDeviceProcAddr(demo->device, "vkGetBufferMemoryRequirements2KHR");
+
+    VkMemoryRequirements2 reqs;
+    memset(&reqs, 0, sizeof(reqs));
+    reqs.sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2_KHR;
+
+    khr_variant(demo->device, &buf_req, &reqs);
+
+    PFN_vkGetBufferMemoryRequirements2 core_variant =
+      (PFN_vkGetBufferMemoryRequirements2)vkGetDeviceProcAddr(demo->device, "vkGetBufferMemoryRequirements2");
+
+    memset(&reqs, 0, sizeof(reqs));
+    reqs.sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2;
+
+    core_variant(demo->device, &buf_req, &reqs);
 }

 static void demo_init_vk_swapchain(struct demo *demo) {
```

The `khr_variant` fetches the requirements normally, while the `core_variant` silently drops the call, fetches no requirements but also doesn't fire any warnings or errors.

**Comment by jzulauf-lunarg** [comment]
@baldurk @mark-lunarg -- probably should review whether RenderDoc and VLV are implementing the same "should return a NULL" policy, as that's something that's been updated.

The layer dispatch tables hold "defensive" noop calls when the downstream returns NULL.  However the tables returning values to gdpa and gipa should not.

**Comment by mark-lunarg** [comment]
Fixed by #1053.
